### DISCUSSION TITLE
PL to PS ISR

### DISCUSSION THREAD

**Discussion by vic9112** [bol-edu/HLS-SOC-Discussions#222] [Q&A] [answered]
We are currently working on a part of our final project. The figure below shows the block diagram of our entire project. To reduce hardware & software overhead, we plan to use the Mailbox (MB) on the FPGA side to generate an interrupt service routine to inform the PS side that our data is ready. However, we are still not sure how to implement this part (FPGA to PS side ISR). I would appreciate any guidance from those who know how to do this.

![ASoC_final](https://github.com/bol-edu/HLS-SOC-Discussions/assets/137171415/c38d697c-c7f2-4177-bbf2-d279d856bf15)

**Comment by Kgfu** [comment, marked as answer]
Hi Vic,
You can't add a new ISR directly in python code.
PYNQ have a guide to introduce how to handle the interrupt which came from PL in PS python code:
https://pynq.readthedocs.io/en/latest/pynq_libraries/interrupt.html
To achieve your requirement, the procedure may need:
1. Add AXI Interrupt Controller in FPGA block design
2. The mailbox (MB) must provide a interrupt pin and connect to the AXI Interrupt Controller
3. Connect the irq of Interrupt Controller to IRQ_F2P[0:0] in PS CPU
Note: Must make sure the property of controller is configured correctly according the guide. Example: Set Interrupt type from Level to Edge.
4. Porting JupyerNotebook Python code to import asyncio library, you should get some example from the guide.

Following images is my experiment before that connect AXI_UART LITE IP to PS via interrupt controller and using Python asyncio library to handler it's interrupt. 
The block design:
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/67a12cb5-0256-415d-9b60-7ef7fa148c00)
Peripheral Interrupt Type in the AXI Interrupt Controller
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/c6246bd6-9e68-44e1-9446-b504e748986b)
Jupyter Notebook:
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/fcffb324-8e47-4d72-9670-dae3bfc2904d)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/bda970c5-3d23-49a5-a465-f908f4448e39)
But the result is always lose the data before the “reg_la0…” string, so I modify the routine as below:
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/04c94474-776b-4ff3-8790-8dd052b72556)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/07069428-5ab9-48e7-a6af-6c1b606af8ed)
Final result is good:
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/5345648/48c0783a-d354-45b5-aa4e-2c3cc450f299)

**Reply by vic9112** [reply]
Thank you for such a detailed and step-by-step complete answer. I will try it !

**Comment by JoshSu0** [comment]
Hi vic9112,

May be the aa_mb_irq can coonect to a input of GPIO controller, and enabling the GPIO input interrupt, the ISR can be hooked to GPIO interrupt.

Below is Xilinx Zynq Vivado GPIO Interrupt Example for reference
After a roughly review. I think it should be help.
https://www.youtube.com/watch?v=JPVTVNtJ7R4

**Reply by vic9112** [reply]
Thank you for providing the method and the video. I will watch and study them!
